nix/shell.go

```diff
@@ -0,0 +1,181 @@
+// Copyright 2022 Jetpack Technologies Inc and contributors. All rights reserved.
+// Use of this source code is governed by the license in the LICENSE file.
+
+package nix
+
+import (
+	"bytes"
+	_ "embed"
+	"fmt"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+	"text/template"
+
+	"github.com/pkg/errors"
+	"go.jetpack.io/devbox/debug"
+)
+
+//go:embed shellrc.tmpl
+var shellrcText string
+var shellrcTmpl = template.Must(template.New("shellrc").Parse(shellrcText))
+
+type name string
+
+const (
+	shUnknown name = ""
+	shBash    name = "bash"
+	shZsh     name = "zsh"
+	shKsh     name = "ksh"
+	shPosix   name = "posix"
+)
+
+// Shell configures a user's shell to run in Devbox. Its zero value is a
+// fallback shell that launches a regular Nix shell.
+type Shell struct {
+	name            name
+	binPath         string
+	userShellrcPath string
+
+	// UserInitHook contains commands that will run at shell startup.
+	UserInitHook string
+}
+
+// DetectShell attempts to determine the user's default shell.
+func DetectShell() (*Shell, error) {
+	path := os.Getenv("SHELL")
+	if path == "" {
+		return nil, errors.New("unable to detect the current shell")
+	}
+
+	sh := &Shell{binPath: filepath.Clean(path)}
+	base := filepath.Base(path)
+	// Login shell
+	if base[0] == '-' {
+		base = base[1:]
+	}
+	switch base {
+	case "bash":
+		sh.name = shBash
+		sh.userShellrcPath = rcfilePath(".bashrc")
+	case "zsh":
+		sh.name = shZsh
+		sh.userShellrcPath = rcfilePath(".zshrc")
+	case "ksh":
+		sh.name = shKsh
+		sh.userShellrcPath = rcfilePath(".kshrc")
+	case "dash", "ash", "sh":
```

Comment:
ah, `dash` is here!

Reply:
Yep! dash and a bunch of shells fall under the "posix" camp where they only support initialization by putting a path to a file in the `ENV` environment variable. That's why there isn't an explicit dash enum.